// homey-api/index.js
/* eslint-disable global-require */

'use strict';

module.exports = {};
module.exports.AthomAppsAPI = require('./lib/AthomAppsAPI');
module.exports.AthomBackupAPI = require('./lib/AthomBackupAPI');
module.exports.AthomCallbackAPI = require('./lib/AthomCallbackAPI');
module.exports.AthomCloudAPI = require('./lib/AthomCloudAPI');
module.exports.AthomConnectAPI = require('./lib/AthomConnectAPI');
module.exports.AthomDNSAPI = require('./lib/AthomDNSAPI');
module.exports.AthomFirmwareAPI = require('./lib/AthomFirmwareAPI');
module.exports.AthomFlowAPI = require('./lib/AthomFlowAPI');
module.exports.AthomGeolocateAPI = require('./lib/AthomGeolocateAPI');
module.exports.AthomIconsAPI = require('./lib/AthomIconsAPI');
module.exports.AthomInfraredAPI = require('./lib/AthomInfraredAPI');
module.exports.AthomNotificationsAPI = require('./lib/AthomNotificationsAPI');
module.exports.AthomSetupAPI = require('./lib/AthomSetupAPI');
module.exports.AthomStoreAPI = require('./lib/AthomStoreAPI');
module.exports.AthomWeatherAPI = require('./lib/AthomWeatherAPI');
module.exports.AthomWebhooksAPI = require('./lib/AthomWebhooksAPI');
module.exports.HomeyCloudAPI = require('./lib/HomeyCloudAPI');
module.exports.HomeyAPI = require('./lib/HomeyAPI/HomeyAPI');
